cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,369 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <rolling/rolling_detail.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/types.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/utilities/nvtx_utils.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/copying.hpp>
+
+#include <rmm/device_scalar.hpp>
+
+#include <memory>
+
+namespace cudf {
+namespace experimental {
+
+namespace detail {
+
+namespace { // anonymous
+
+/**
+ * @brief Computes the rolling window function
+ *
+ * @tparam ColumnType  Datatype of values pointed to by the pointers
+ * @tparam agg_op  A functor that defines the aggregation operation
+ * @tparam is_mean Compute mean=sum/count across all valid elements in the window
+ * @tparam block_size CUDA block size for the kernel
+ * @tparam WindowIterator iterator type (inferred)
+ * @param input Input column device view
+ * @param output Output column device view
+ * @param window_begin[in] Rolling window size iterator, accumulates from
+ *                in_col[i-window+1] to in_col[i] inclusive
+ * @param forward_window_begin[in] Rolling window size iterator in the forward
+ *                direction, accumulates from in_col[i] to
+ *                in_col[i+forward_window] inclusive
+ * @param min_periods[in]  Minimum number of observations in window required to
+ *                have a value, otherwise 0 is stored in the valid bit mask
+ */
+template <typename T, typename agg_op, rolling_operator op, int block_size, bool has_nulls,
+          typename WindowIterator>
+__launch_bounds__(block_size)
+__global__
+void gpu_rolling(column_device_view input,
+                 mutable_column_device_view output,
+                 size_type * __restrict__ output_valid_count,
+                 WindowIterator window_begin,
+                 WindowIterator forward_window_begin,
+                 size_type min_periods)
+{
+  size_type i = blockIdx.x * block_size + threadIdx.x;
+  size_type stride = block_size * gridDim.x;
```

Comment:
These should be initialised closer to where they are used aka below:
```cpp
    agg_op agg_operator{};
  gdf_size_type warp_valid_count{0};
```

Reply:
warp_valid_count must be initialized outside the loop. I figure the agg_op struct initialization will get optimized away, so I can move it.